### PR TITLE
add support for new output option jsonpScriptType

### PR DESCRIPTION
**Pull request by neeharv**
**What kind of change does this PR introduce?**

This PR introduces a new output option `jsonpScriptType`, which allows us to customize the `type` attribute of the script tag the webpack runtime adds to the DOM. 

With this PR, we can add `outputOptions : { jsonpScriptType : 'module' }` to have the runtime load async chunks using `<script type="module" src="myAsyncChunk.js" />`. The default option is `text/javascript`, so this is fully backwards compatible.

**Did you add tests for your changes?**

I couldn't find any tests for `JsonpMainTemplatePlugin.js`, any guidance on where/how to add integration tests for this particular feature would be helpful and I'll do the same.

**If relevant, link to documentation update:**

Docs PR - https://github.com/webpack/webpack.js.org/pull/1770

**Summary**

This option is needed for bundling ES6 ready code and shipping it for modern browsers using `<script type="module">` as detailed [here](https://philipwalton.com/articles/deploying-es2015-code-in-production-today/).

In situations where we have code split on routes, and we know what is the async chunk of the current page (so it can be preloaded / downloaded ahead of time), we can add a `<script type="module" src="myAsyncChunk.js"/>` to the head of the page, but when the webpack runtime loads, it downloads the async chunk using `<script type="text/javascript" src="myAsyncChunk.js" />`. This causes browsers to re-downloaded the resource as the two are treated differently.

**Does this PR introduce a breaking change?**

This change is fully backwards compatible

**Other information**

### CODE REVIEW

File: lib/JsonpMainTemplatePlugin.js [hunk 1]
```diff
@@ -48,7 +49,7 @@ class JsonpMainTemplatePlugin {
 			});
 			return this.asString([
 				"var script = document.createElement('script');",
-				"script.type = 'text/javascript';",
+				`script.type = '${jsonpScriptType}';`,
```
Comment (sokra) [inline lib/JsonpMainTemplatePlugin.js]: `` `script.type = ${JSON.stringify(jsonpScriptType};`, ``

File: lib/JsonpMainTemplatePlugin.js [hunk 2]
```diff
@@ -29,6 +29,7 @@ class JsonpMainTemplatePlugin {
 			const chunkMaps = chunk.getChunkMaps();
 			const crossOriginLoading = this.outputOptions.crossOriginLoading;
 			const chunkLoadTimeout = this.outputOptions.chunkLoadTimeout;
+			const jsonpScriptType = this.outputOptions.jsonpScriptType || "text/javascript";
```
Comment (sokra) [inline lib/JsonpMainTemplatePlugin.js]: remove the default here. add the default in `lib/WebpackOptionsDefaulter.js`

**Comment by sokra** [comment]
Thanks

**Comment by code-smith** [comment]
@neeharv  Awesome!!

File: lib/WebpackOptionsDefaulter.js [hunk 3]
```diff
@@ -61,6 +61,7 @@ class WebpackOptionsDefaulter extends OptionsDefaulter {
 		this.set("output.hotUpdateChunkFilename", "[id].[hash].hot-update.js");
 		this.set("output.hotUpdateMainFilename", "[hash].hot-update.json");
 		this.set("output.crossOriginLoading", false);
+		this.set("output.jsonpScriptType", "text/javascript");
```
Comment (mathiasbynens) [inline lib/WebpackOptionsDefaulter.js]: `type="text/javascript"` is [the implicit default](https://mathiasbynens.be/notes/html5-levels#type-attributes) for all `<script>`s, and should thus be omitted from the output.  Ref. #5221. Does anyone want to submit a follow-up PR?

Comment (mathiasbynens) [reply lib/WebpackOptionsDefaulter.js]: i.e. we should either output `<script type="module" …>` or just `<script>`.

**Pull request merged**